README.md
## connectpy

Flask Client-Server Connect5 Game 

To run connectpy_server (0.0.0.0:80 default):
* $ git clone git@github.com:gaffer-93/connectpy.git
* $ cd connectpy
* $ make install
* $ ./bin/run_connectpy_server

To run connectpy_client:
* $ python src/connectpy_client.py (localhost:80 default)

To run tests:
* $ make test

To check out the CircleCI build history:
* Go to https://circleci.com/gh/gaffer-93/connectpy 

To grab the latest .deb package from CircleCI:
* Go to https://19-215333119-gh.circle-artifacts.com/0/tmp/artifacts/connectpy_19%2Bmaster_all.deb


debian/changelog
connectpy (1.0.0) unstable; urgency=low

  * Initial release.

 -- connectpy Package Maintainer <cgaffers@gmail.com>  Tue, 15 Oct 2019 14:57:37 +0000

tests/test_connectpy.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import flask_testing
import connectpy_server
import connectpy_game
import unittest
import os
import mock


class TestConnectpyServer(flask_testing.TestCase):

    def create_app(self):
        self.dir = os.path.dirname(os.path.abspath(__file__))
        os.environ['CONNECTPY_SETTINGS'] = os.path.join(self.dir, 'test.cfg')
        return connectpy_server.create_app()

    def setUp(self):
        self.mock_response = mock.Mock()
        self.mock_game_obj = mock.Mock()
        self.app.game = self.mock_game_obj
        self.player_id = 'deadbeef'
        self.client = self.app.test_client()

    def _test_not_joined(self, endpoint):
        data = {'player_id': self.player_id, 'column': 1}
        self.app.game.get_player_indicator.side_effect = \
            connectpy_server.conn_py.PlayerInvalidException
        rv = self.client.post(endpoint, json=data)
        self.assertEqual(rv.status_code, 403)

    def _test_bad_mimetype(self, endpoint):
        data = {'player_id': self.player_id}
        rv = self.client.post(endpoint, data=data)
        self.assertEqual(rv.status_code, 415)

    def _test_required_fields(self, endpoint, data):
        rv = self.client.post(endpoint, json=data)
        self.assertEqual(rv.status_code, 400)

    def test_status_player_not_joined(self):
        self._test_not_joined('/status')

    def test_status_bad_mimetype(self):
        self._test_bad_mimetype('/status')

    def test_status_required_fields(self):
        self._test_required_fields('/status', {})

    def test_status_ok(self):
        test_data = {'test': 'ok'}
        self.app.game.dict = test_data

        rv = self.client.post('/status', json={'player_id': self.player_id})
        self.assertEqual(rv.status_code, 200)
        self.assertEqual(rv.json, test_data)

    def test_join_required_fields(self):
        self._test_required_fields('/join', {})

    def test_join_bad_mimetype(self):
        self._test_bad_mimetype('/join')

    def test_join_game_started(self):
        self.app.game.closed = False
        self.app.game.started = True
        self.app.game.players_ready = True

        rv = self.client.post('/join', json={'player_id': self.player_id})
        self.assertEqual(rv.status_code, 503)

    def test_join_game_already_joined(self):
        self.app.game.closed = False
        self.app.game.started = False
        self.app.game.players_ready = False
        self.app.game.add_player.side_effect = \
            connectpy_server.conn_py.AlreadyJoinedException

        rv = self.client.post('/join', json={'player_id': self.player_id})
        self.assertEqual(rv.status_code, 409)

    def test_join_game_ok(self):
        self.app.game.closed = False
        self.app.game.started = False
        self.app.game.players_ready = False
        test_data = {'test': 'ok'}
        self.app.game.dict = test_data

        rv = self.client.post('/join', json={'player_id': self.player_id})
        self.assertEqual(rv.status_code, 200)
        self.assertEqual(rv.json, test_data)

    def test_move_player_not_joined(self):
        self._test_not_joined('/move')

    def test_move_bad_mimetype(self):
        self._test_bad_mimetype('/move')

    def test_move_required_fields(self):
        self._test_required_fields('/move', {})

    def test_move_not_your_turn(self):
        self.app.game.is_turn.return_value = False

        rv = self.client.post(
            '/move', json={'player_id': self.player_id, 'column': 1})
        self.assertEqual(rv.status_code, 420)

    def test_move_invalid(self):
        self.app.game.is_turn.return_value = True

        # Test FullColumnException
        self.app.game.drop_disc.side_effect = \
            connectpy_server.conn_py.FullColumnException
        rv = self.client.post(
            '/move', json={'player_id': self.player_id, 'column': 1})
        self.assertEqual(rv.status_code, 400)

        # Test ColumnOutOfBoundsException
        self.app.game.drop_disc.side_effect = \
            connectpy_server.conn_py.ColumnOutOfBoundsException
        rv = self.client.post(
            '/move', json={'player_id': self.player_id, 'column': 1})
        self.assertEqual(rv.status_code, 400)

    def test_move_ok(self):
        self.app.game.is_turn.return_value = True
        test_data = {'test': 'ok'}
        self.app.game.dict = test_data

        # Test normal move
        self.app.game.drop_disc.return_value = False
        rv = self.client.post(
            '/move', json={'player_id': self.player_id, 'column': 1})
        self.assertEqual(rv.status_code, 200)
        self.assertEqual(rv.json, test_data)

        # Test winning move
        self.app.game.drop_disc.return_value = True
        rv = self.client.post(
            '/move', json={'player_id': self.player_id, 'column': 1})
        self.assertEqual(rv.status_code, 200)
        self.app.game.reset_game.assert_called_once()

    def test_close_player_not_joined(self):
        self._test_not_joined('/close')

    def test_close_bad_mimetype(self):
        self._test_bad_mimetype('/close')

    def test_close_required_fields(self):
        self._test_required_fields('/close', {})

    def test_close_ok(self):
        test_data = {'test': 'ok'}
        self.app.game.dict = test_data

        rv = self.client.post('/close', json={'player_id': self.player_id})
        self.assertEqual(rv.status_code, 200)
        self.assertEqual(rv.json, test_data)


class TestConnectpyGame(unittest.TestCase):

    def setUp(self):
        self.config = {
            "game_columns": 9,
            "game_rows": 6,
            "win_zone": 5
        }
        self.game = connectpy_game.ConnectPyGame(self.config)

    def test_players_ready(self):
        # No players
        self.assertFalse(self.game.players_ready)
        # 1 player
        self.game.players = {"a": 1}
        self.assertFalse(self.game.players_ready)
        # 2 players
        self.game.players["b"] = 2
        self.assertTrue(self.game.players_ready)

    def test_dict(self):
        expected = {
            "game": self.game.grid,
            "turn": self.game.current_turn,
            "players": self.game.players,
            "winner": self.game.winner,
            "started": self.game.started,
            "last_drop": self.game.last_drop,
            "rows": self.game.rows,
            "columns": self.game.columns,
            "closed": self.game.closed
        }
        self.assertEqual(expected, self.game.dict)

    def test_start_game_players_not_ready(self):
        # No players
        with self.assertRaises(connectpy_game.PlayersNotReadyException):
            self.game.start_game()

    def test_start_game_ok(self):
        # Test not initialised
        self.assertFalse(self.game.grid)
        self.assertFalse(self.game.current_turn)
        self.assertFalse(self.game.started)

        # Players ready
        self.game.players = {"a": 1, "b": 2}
        self.game.start_game()
        self.assertTrue(self.game.grid)
        self.assertTrue(self.game.current_turn in self.game.players)
        self.assertTrue(self.game.started)

    def test_reset_grid(self):
        self.game.grid = [['test']]
        self.game.last_drop = (1, 2)
        self.game.winner = "a"
        self.game.reset_game()

        self.assertTrue(
            all([all([n == 0 for n in row]) for row in self.game.grid]))
        self.assertFalse(self.game.last_drop)
        self.assertFalse(self.game.winner)

    def test_drop_disc_bad_move(self):
        self.game.players = {"a": 1, "b": 2}
        self.game.start_game()

        with self.assertRaises(connectpy_game.FullColumnException):
            for _ in range(self.game.rows + 1):
                self.game.drop_disc("a", 1)

        with self.assertRaises(connectpy_game.ColumnOutOfBoundsException):
            self.game.drop_disc("a", self.game.columns + 1)

    def test_drop_disc_ok(self):
        self.game.players = {"a": 1, "b": 2}
        self.game.start_game()

        for i in range(self.game.columns):
            # Drop player's discs and check turn is handed over
            self.game.drop_disc(self.game.current_turn, i)
            last_turn = self.game.current_turn
            last_drop = self.game.last_drop

            self.game.drop_disc(self.game.current_turn, i)
            self.assertFalse(self.game.current_turn == last_turn)
            self.assertFalse(self.game.last_drop == last_drop)

        self.assertEqual(self.game.winner, "b")

        # Test expected game state - last 2 rows filled with respective player
        # indicators
        empty_row = [0 for _ in range(self.game.columns)]
        expected_state = [empty_row for row in range(self.game.rows)]
        expected_state[-1] = [1 for _ in range(self.game.columns)]
        expected_state[-2] = [2 for _ in range(self.game.columns)]
        self.assertEqual(self.game.grid, expected_state)

    def test_axis_has_winner(self):
        self.game.win_zone = 5
        win_axis = [1, 1, 1, 0, 0, 0, 0, 0, 0]
        self.assertFalse(self.game.axis_has_winner(1, win_axis))
        win_axis = [1, 1, 1, 1, 1, 0, 0, 0, 0]
        self.assertTrue(self.game.axis_has_winner(1, win_axis))

    def test_is_winner(self):
        self.game.win_zone = 5

        no_win_state = [
            [0, 0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 1, 0, 0, 0, 0, 0, 0],
            [0, 0, 2, 0, 0, 0, 0, 0, 0],
            [0, 0, 2, 1, 2, 0, 0, 0, 0],
            [0, 0, 2, 2, 2, 1, 1, 0, 0],
            [0, 0, 2, 1, 1, 1, 1, 0, 0]]
        drop_coords = [4, 6]
        self.game.grid = no_win_state
        self.assertFalse(self.game.is_winner(1, drop_coords))

        horizontal_win_state = [
            [0, 0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 1, 0, 0, 0, 0, 0, 0],
            [0, 0, 2, 0, 0, 0, 0, 0, 0],
            [0, 0, 2, 1, 2, 0, 0, 0, 0],
            [0, 0, 2, 2, 2, 1, 0, 0, 0],
            [0, 0, 2, 1, 1, 1, 1, 1, 0]]
        drop_coords = [5, 7]
        self.game.grid = horizontal_win_state
        self.assertTrue(self.game.is_winner(1, drop_coords))

        vertical_win_state = [
            [0, 0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 2, 0, 0, 0, 0, 0, 0],
            [0, 0, 2, 0, 0, 0, 0, 0, 0],
            [0, 0, 2, 1, 0, 1, 0, 0, 0],
            [0, 0, 2, 2, 2, 1, 1, 0, 0],
            [0, 0, 2, 1, 1, 1, 2, 0, 0]
        ]
        drop_coords = [1, 2]
        self.game.grid = vertical_win_state
        self.assertTrue(self.game.is_winner(2, drop_coords))

        diagonal_win_state = [
            [0, 0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 1, 0, 0, 0, 2, 0, 0],
            [0, 0, 2, 0, 0, 2, 2, 0, 0],
            [0, 0, 2, 1, 2, 1, 1, 0, 0],
            [0, 0, 2, 2, 2, 1, 1, 0, 0],
            [0, 1, 2, 1, 1, 1, 2, 0, 0]
        ]
        drop_coords = [1, 6]
        self.game.grid = diagonal_win_state
        self.assertTrue(self.game.is_winner(2, drop_coords))

        diagonal_flip_win_state = [
            [0, 0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 1, 1, 0, 0, 1, 0, 0],
            [0, 0, 2, 2, 1, 2, 2, 0, 0],
            [0, 2, 2, 1, 2, 1, 1, 0, 0],
            [0, 2, 2, 2, 2, 1, 1, 0, 0],
            [0, 1, 2, 1, 1, 1, 2, 1, 0]
        ]
        drop_coords = [5, 7]
        self.game.grid = diagonal_flip_win_state
        self.assertTrue(self.game.is_winner(1, drop_coords))


    def test_add_player(self):
        self.game.add_player("a")
        self.assertFalse(self.game.players_ready)
        self.assertEqual(self.game.players["a"], 1)
        self.game.add_player("b")
        self.assertTrue(self.game.players_ready)
        self.assertEqual(self.game.players["b"], 2)

    def test_add_player_already_joined(self):
        self.game.players = {"a": 1}
        with self.assertRaises(connectpy_game.AlreadyJoinedException):
            self.game.add_player("a")

    def test_add_player_full_game(self):
        self.game.players = {"a": 1, "b": 2}
        with self.assertRaises(connectpy_game.FullGameException):
            self.game.add_player("a")

    def test_is_turn(self):
        self.game.current_turn = "a"
        self.assertFalse(self.game.is_turn("b"))
        self.assertTrue(self.game.is_turn("a"))

    def test_next_player(self):
        self.game.players = {"a": 1, "b": 2}
        self.game.start_game()

        self.assertEqual(self.game.next_player(), "b")
        self.assertEqual(self.game.next_player(), "a")

    def test_get_player_indicator_not_joined(self):
        self.game.players = {"a": 1, "b": 2}
        with self.assertRaises(connectpy_game.PlayerInvalidException):
            self.game.get_player_indicator("c")

    def test_get_player_indicator(self):
        self.game.players = {"a": 1, "b": 2}
        self.assertEqual(
            self.game.get_player_indicator("a"),
            self.game.players["a"]
        )

    def test_close(self):
        self.game.players = {"a": 1, "b": 2}
        self.game.start_game()
        self.game.close("a")

        self.assertEqual(self.game.closed, "a")



src/connectpy_game.py
# -*- coding: utf-8 -*-

from itertools import islice, cycle


class ColumnOutOfBoundsException(Exception):
    pass


class FullColumnException(Exception):
    pass


class FullGameException(Exception):
    pass


class PlayersNotReadyException(Exception):
    pass


class PlayerInvalidException(Exception):
    pass


class AlreadyJoinedException(Exception):
    pass


def window(seq, n):
    """
    Returns a sliding window (of width `n`) over data from the iterable `seq`
       s -> (s0,s1,...s[n-1]), (s1,s2,...,sn), ...
    https://docs.python.org/release/2.3.5/lib/itertools-example.html
    """
    it = iter(seq)
    result = tuple(islice(it, n))
    if len(result) == n:
        yield result
    for elem in it:
        result = result[1:] + (elem,)
        yield result


def surrounding_slice(arr, idx, n):
    """
    Returns a surrounding slice (of width `n`) from the iterable `arr` at
    index `idx`
    """
    return arr[max(idx - n, 0):min(idx + n, len(arr))]


def surrounding_diag(mat, x, y, n, flip=False):
    """
    Returns a surrounding diagonal slice (of width `n`) from a matrix `mat` at
    coordinates `x`, `y`, setting `flip` returns the opposite diagonal
    """
    diag = []
    direction = 1 if flip else -1
    for i in range(-n, n):
        xi, yi = x - i, y + (i * direction)
        try:
            if xi * yi >= 0:
                diag.append(mat[xi][yi])
        except IndexError:
            continue
    return diag


class ConnectPyGame(object):
    """Object for storing and updating ConnectPy game state."""

    def __init__(self, config):
        """
        Expect a `config` of the form:
            game_columns: 9
            game_rows: 6
            win_zone: 5
        """
        self.config = config
        self.columns = config.get('game_columns', 9)
        self.rows = config.get('game_rows', 6)
        self.win_zone = config.get('win_zone', 5)
        self.started = False
        self.max_players = 2
        self.players = {}
        self.player_cycle = None
        self.current_turn = None
        self.winner = None
        self.last_drop = None
        self.closed = False
        self.grid = []

    @property
    def players_ready(self):
        """Returns a bool indicating if enough players have joined"""
        return len(self.players) == self.max_players

    @property
    def dict(self):
        """Returns a dict representation of the ConnectPyGame object"""
        return {
            "game": self.grid,
            "turn": self.current_turn,
            "players": self.players,
            "winner": self.winner,
            "started": self.started,
            "last_drop": self.last_drop,
            "rows": self.rows,
            "columns": self.columns,
            "closed": self.closed
        }

    def start_game(self):
        """
        If all players have joined, set the `self.grid` and pick the first
        player to move `self.current_turn`
        """
        if self.players_ready:
            self.reset_game()
            self.player_cycle = cycle(self.players.keys())
            self.current_turn = self.next_player()
            self.started = True
        else:
            raise PlayersNotReadyException(
                "Calling start_game before all players have connected")

    def reset_game(self):
        """Reset game state to starting state"""
        self.grid = []
        for row in range(self.rows):
            self.grid.append([0 for column in range(self.columns)])
        self.last_drop = None
        self.winner = None

    def drop_disc(self, player_id, column_idx):
        """
        Returns True if the drop move for `player_id` at `column_idx` is a
        winning move. Also cycles `self.current_turn` to the next player
        """
        player_indicator = self.get_player_indicator(player_id)

        drop_coords = None
        try:
            for row_idx, row in reversed(list(enumerate(self.grid))):
                if row[column_idx] == 0:
                    row[column_idx] = player_indicator
                    drop_coords = (row_idx, column_idx)
                    break
            else:
                raise FullColumnException(
                    "Player {} - Column {} full".format(
                        player_id, column_idx))
        except IndexError:
            raise ColumnOutOfBoundsException(
                "Player {} - Column {} out of bounds".format(
                    player_id, column_idx))

        self.current_turn = self.next_player()
        self.last_drop = drop_coords

        is_won = self.is_winner(player_indicator, drop_coords)
        if is_won:
            self.winner = player_id

        return is_won


    def axis_has_winner(self, player, win_axis):
        """
        Returns True if the list `win_axis` contains a chain of indicators of
        length `self.win_zone` for `player`
        """
        for seq in window(win_axis, self.win_zone):
            if all(n == player for n in seq):
                return True

    def is_winner(self, player, drop_coords):
        """
        Returns True if the matrix `self.grid` contains a chain of indicators
        of length `self.win_zone` for `player` in any direction from
        coordinates `drop_coords`.
        """
        row_idx, column_idx = drop_coords

        # Get the surrounding win zone on the horizontal axis
        win_hor = surrounding_slice(
            self.grid[row_idx], column_idx, self.win_zone)

        # Get the surrounding win zone on the vertical axis
        win_vert = surrounding_slice(
            list(zip(*self.grid))[column_idx], row_idx, self.win_zone)

        # Get the surrounding win zone on the diagonal axis
        win_diag_main = surrounding_diag(
            self.grid, row_idx, column_idx, self.win_zone, flip=False)

        # Get the surrounding win zone on the flipped diagonal axis
        win_diag_flip = surrounding_diag(
            self.grid, row_idx, column_idx, self.win_zone, flip=True)

        # Return true is any win zone axis contains a winning chain
        return any(self.axis_has_winner(player, win_axis) for win_axis in
                   [win_hor, win_vert, win_diag_main, win_diag_flip])

    def add_player(self, player_id):
        """Adds a player_id to `self.players` and assigns it an indicator"""
        if not self.players_ready:
            if player_id not in self.players:
                self.players[player_id] = len(self.players) + 1
            else:
                raise AlreadyJoinedException(
                    "Player {} already joined".format(player_id))
        else:
            raise FullGameException("Maximum players reached")

    def is_turn(self, player_id):
        """Returns True if `player_id` matches `self.current_turn`"""
        return self.current_turn == player_id

    def next_player(self):
        """Returns the next player_id from `self.player_cycle`"""
        return self.player_cycle.__next__()

    def get_player_indicator(self, player_id):
        """Returns the player indicator for `player_id`"""
        try:
            return self.players[player_id]
        except KeyError:
            raise PlayerInvalidException(
                "Player ID {} not joined".format(player_id))

    def close(self, player_id):
        """Sets `self.closed` to `player_id`"""
        self.closed = player_id



src/connectpy_server.py
# -*- coding: utf-8 -*-

import yaml
import os
import connectpy.connectpy_game as conn_py

from functools import wraps
from flask import Flask, Blueprint, request, jsonify, current_app

paths = Blueprint('paths', __name__)


def game_started(func):
    @wraps(func)
    def decorator(*args, **kwargs):
        if current_app.game.closed:
            new_game(current_app)
        if current_app.game.started or current_app.game.players_ready:
            return error_response(
                "ConnectPy game in progress", status=503)
        else:
            resp = func(*args, **kwargs)
        return resp
    return decorator


def player_joined(func):
    @wraps(func)
    def decorator(*args, **kwargs):
        try:
            player_id = request.json.get('player_id')
            current_app.game.get_player_indicator(player_id)
            request.player_id = player_id
        except conn_py.PlayerInvalidException as e:
            return error_response(str(e), status=403)
        else:
            resp = func(*args, **kwargs)
        return resp
    return decorator


def required_fields(fields):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            missing = []
            if request.mimetype == 'application/json':
                for field in fields:
                    try:
                        setattr(request, field, request.json[field])
                    except KeyError:
                        missing.append(field)
                if missing:
                    return error_response(
                        "{} field(s) required".format(','.join(missing)),
                        status=400)
            else:
                return error_response(
                    "Unsupported Content-Type - expected: application/json",
                    status=415)
            resp = func(*args, **kwargs)
            return resp
        return wrapper
    return decorator


@paths.route('/join', methods=['POST'])
@required_fields(['player_id'])
@game_started
def join():
    try:
        current_app.game.add_player(request.player_id)
    except conn_py.AlreadyJoinedException as e:
        return error_response(str(e), status=409)
    else:
        print("Player {} connected".format(request.player_id))

    if current_app.game.players_ready:
        current_app.game.start_game()
        print("Game started")

    return ok_response(current_app.game.dict)


@paths.route('/status', methods=['GET', 'POST'])
@required_fields(['player_id'])
@player_joined
def status():
    return ok_response(current_app.game.dict)


@paths.route('/move', methods=['POST'])
@required_fields(['player_id', 'column'])
@player_joined
def move():
    player_id = request.json['player_id']
    column = request.json['column']

    if current_app.game.is_turn(player_id):
        try:
            winner = current_app.game.drop_disc(player_id, column)
        except (conn_py.FullColumnException,
                conn_py.ColumnOutOfBoundsException) as e:
            return error_response(str(e), status=400)

        game_dict = current_app.game.dict
        current_app.game.print_grid()
        if winner:
            print("{} Wins! - Resetting".format(player_id))
            current_app.game.reset_game()
        return ok_response(game_dict)
    else:
        return error_response(
            "Not your turn! - Enhance your calm", status=420)


@paths.route('/close', methods=['POST'])
@required_fields(['player_id'])
@player_joined
def close():
    current_app.game.close(request.player_id)
    resp = ok_response(current_app.game.dict)
    print("Game closed by {}".format(request.player_id))

    return resp


def get_config():
    config_filename = os.environ.get('CONNECTPY_SETTINGS')
    config = {}
    if config_filename:
        with open(config_filename) as f:
            config.update(yaml.load(f))
    else:
        print("No config specified, using defaults")
    return config


def error_response(error_str, status=400):
    resp = jsonify({'error': error_str})
    resp.status_code = status
    return resp


def ok_response(data):
    resp = jsonify(data)
    resp.status_code = 200
    return resp


def new_game(app):
    app.game = conn_py.ConnectPyGame(app.config)


def create_app():
    config = get_config()
    app = Flask(__name__)
    app.config.update(config)
    app.register_blueprint(paths)
    new_game(app)

    return app


if __name__ == '__main__':
    create_app()


src/connectpy_client.py
# -*- coding: utf-8 -*-

import requests
import argparse
import time
import sys
import signal
import yaml

from termios import tcflush, TCIFLUSH


class PlayerClient(object):
    def __init__(self, player_id, server_url):
        self.game_state = {}
        self.last_game_state = {}
        self.id = player_id
        self.server_url = server_url

    @property
    def opposing_player(self):
        for player_id, indicator in self.game_state['players'].items():
            if player_id != self.id:
                return player_id
        else:
            return None

    @property
    def can_move(self):
        return self.game_state['turn'] == self.id

    def make_request(self, endpoint, data):
        resp = requests.post(
            self.server_url + endpoint, json=data)
        if resp:
            self.last_game_state = self.game_state
            self.game_state = resp.json()

        return resp

    def join_server(self):
        return self.make_request('/join', {'player_id': self.id})

    def update_status(self):
        return self.make_request('/status', {'player_id': self.id})

    def make_move(self, column):
        return self.make_request(
            '/move', {'player_id': self.id, 'column': column})

    def close_game(self):
        return self.make_request(
            '/close', {'player_id': self.id})

    def wait_for_opponent(self):
        while not self.opposing_player:
            self.update_status()
            time.sleep(1)

    def printable_state(self):
        s = [['[   ]' if e == 0
             else '[ {} ]'.format(play_piece(e)) for e in row]
             for row in self.game_state['game']]
        s.append(['[ {} ]'.format(n + 1) for n in range(
            self.game_state['columns'])])
        lens = [max(map(len, col)) for col in zip(*s)]
        fmt = ' '.join('{{:{}}}'.format(x) for x in lens)
        table = [fmt.format(*row) for row in s]
        return '\n'.join(table)

    def get_move(self):
        column = None
        while not column:
            try:
                tcflush(sys.stdin, TCIFLUSH)
                number = int(input(
                    "Pick a column ^, or press 0 to end game: "))
                if number == 0:
                    self.close_game()
                    column = number
                    break
                elif not 1 <= number <= self.game_state['columns']:
                    print('Out of bounds! - Pick again...')
                else:
                    column = number
            except ValueError:
                print('Not an integer! - Pick again...')

        return column - 1


def play_piece(player_indicator):
        return 'x' if player_indicator == 1 else 'o'


def get_player_client(server_url):
    player_id = input("Please enter your name: ")
    player_client = PlayerClient(player_id, server_url)

    return player_client


def try_join_game(server_url):
    player_client = get_player_client(server_url)
    print("Hi {}! - Connecting to game...".format(player_client.id))
    resp = player_client.join_server()

    if resp.status_code == 409:
        print(resp.json()['error'])
        new_player_id = None
        while new_player_id == player_client.id:
            new_player_id = input("Please enter a different name: ")
        player_client.id = new_player_id
        resp = player_client.join_server()

    if resp:
        print("Player {} joined successfully!".format(player_client.id))
        print("Waiting for opponent...")
        player_client.wait_for_opponent()
        print("Your opponent is {}, good luck!".format(
            player_client.opposing_player))
        print(player_client.printable_state())
        return player_client
    else:
        print(resp.json()['error'])
        return False


def print_state_change(player_client):
    if player_client.game_state != player_client.last_game_state:
        print(player_client.printable_state())
    else:
        print('Waiting for opponent...\r', end="")


def run_client(
        server_url='http://localhost:80', interval=0.5, wait_timeout=30):
    player_client = None
    while not player_client:
        player_client = try_join_game(server_url)

    time_waiting = 0

    while True:
        if player_client.can_move:
            time_waiting = 0
            column = player_client.get_move()
            resp = player_client.make_move(column)
            if not resp:
                print(resp.json()['error'])
        else:
            time_waiting += interval
            if time_waiting >= wait_timeout:
                player_client.close_game()
                print("Closing game due to inactiviy...")
            player_client.update_status()

        winner = player_client.game_state['winner']
        if winner:
            print("{} Wins! - Resetting".format(winner))

        player_closed = player_client.game_state['closed']
        if player_closed:
            print("Game closed by {}, goodbye!".format(player_closed))
            break

        print_state_change(player_client)
        time.sleep(interval)


def get_config(config_path):
    config = {}
    if config_path:
        with open(config_path) as f:
            config.update(yaml.load(f))
    else:
        print("No config specified, defaults will be used")
    return config


def main():
    parser = argparse.ArgumentParser(description='ConnectPy user client')
    parser.add_argument(
        '-c',
        dest='config_path',
        action='store',
        help='ConnectPy client config'
    )
    args = parser.parse_args()
    config = get_config(args.config_path)
    run_client(**config)

if __name__ == '__main__':
    main()
